File: app/baseline_agent.ts
/**
 * @license
 * SPDX-License-Identifier: Apache-2.0
 */

import { State, Action, ActionType, MachineStatus } from "./models.ts";

export class BaselineAgent {
  public static getAction(state: State): Action {
    // 1. Prioritize maintenance for broken machines
    const brokenMachine = state.machines.find(m => m.status === MachineStatus.BROKEN);
    if (brokenMachine) {
      return { type: ActionType.MAINTENANCE, machine_id: brokenMachine.id };
    }

    // 2. Proactive maintenance for low health machines (only if idle)
    const lowHealthMachine = state.machines.find(
      m => m.health < 0.4 && m.status === MachineStatus.IDLE
    );
    if (lowHealthMachine) {
      return { type: ActionType.MAINTENANCE, machine_id: lowHealthMachine.id };
    }

    // 3. Assign jobs to idle machines
    const idleMachine = state.machines.find(m => m.status === MachineStatus.IDLE);
    if (idleMachine && state.job_queue.length > 0) {
      // Priority-aware: pick highest priority job, then longest processing time
      const sortedJobs = [...state.job_queue].sort((a, b) => {
        if (b.priority !== a.priority) return b.priority - a.priority;
        return b.processing_time - a.processing_time;
      });
      
      const bestJob = sortedJobs[0];
      
      // Check energy constraint before assigning
      const estimatedNewEnergy = state.current_energy_usage + (10.0 + bestJob.priority * 2.0);
      if (estimatedNewEnergy <= state.energy_budget * 1.1) { // Allow slight overage for high priority
        return { 
          type: ActionType.ASSIGN, 
          job_id: bestJob.id, 
          machine_id: idleMachine.id 
        };
      }
    }

    return { type: ActionType.DO_NOTHING };
  }
}

File: agents/baseline_agent.py
import requests
import os

class BaselineAgent:
    def __init__(self, env_url="http://localhost:3000"):
        self.env_url = env_url

    def get_action(self, state):
        """
        Simple rule-based logic to decide the next action.
        """
        # 1. Check for broken machines
        for machine in state["machines"]:
            if machine["status"] == "broken":
                return {"type": "maintenance", "machine_id": machine["id"]}
        
        # 2. Check for low health machines
        for machine in state["machines"]:
            if machine["health"] < 0.5 and machine["status"] == "idle":
                return {"type": "maintenance", "machine_id": machine["id"]}
        
        # 3. Assign jobs to idle machines
        idle_machines = [m for m in state["machines"] if m["status"] == "idle"]
        if idle_machines and state["job_queue"]:
            # Priority-aware: pick highest priority job
            best_job = sorted(state["job_queue"], key=lambda j: j["priority"], reverse=True)[0]
            return {
                "type": "assign", 
                "job_id": best_job["id"], 
                "machine_id": idle_machines[0]["id"]
            }
        
        return {"type": "do_nothing"}

    def run_episode(self, task_id="easy"):
        res = requests.post(f"{self.env_url}/reset", json={"task_id": task_id})
        state = res.json()
        
        while not state.get("done", False):
            action = self.get_action(state)
            res = requests.post(f"{self.env_url}/step", json={"action": action})
            state = res.json()["state"]
        
        res = requests.get(f"{self.env_url}/grader")
        return res.json()["score"]

if __name__ == "__main__":
    agent = BaselineAgent()
    score = agent.run_episode("easy")
    print(f"Baseline Agent Score (Easy): {score}")
